modin/core/dataframe/base/dataframe/dataframe.py

```diff
@@ -0,0 +1,506 @@
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+"""
+Module contains class ModinDataframe.
+
+ModinDataframe is a parent abstract class for any dataframe class.
+"""
+
+from abc import ABC, abstractmethod
+from typing import List, Hashable, Optional, Callable, Union, Dict
+
+
+class ModinDataframe(ABC):
+    """
+    An abstract class that represents the Parent class for any DataFrame class.
+
+    This class is intended to specify the behaviours that a DataFrame must implement.
+
+    For more details about how these methods were chosen, please refer to this
+    (http://www.vldb.org/pvldb/vol13/p2033-petersohn.pdf) paper, which specifies
+    a DataFrame algebra that this class exposes.
+    """
+
+    @abstractmethod
+    def mask(
+        self,
+        row_labels: Optional[List[Hashable]] = None,
+        row_positions: Optional[List[int]] = None,
+        col_labels: Optional[List[Hashable]] = None,
+        col_positions: Optional[List[int]] = None,
+    ) -> "ModinDataframe":
+        """
+        Mask rows and columns in the dataframe.
+
+        Allow users to perform selection and projection on the row and column number (positional notation),
+        in addition to the row and column labels (named notation)
+
+        Parameters
+        ----------
+        row_labels : list of hashable, default: None
+            The row labels to extract.
+        row_positions : list of int, default: None
+            The row indices to extract.
+        col_labels : list of hashable, default: None
+            The column labels to extract.
+        col_positions : list of int, default: None
+            The column indices to extract.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the mask provided.
+        """
+        pass
+
+    @abstractmethod
+    def filter_by_types(self, types: List[Hashable]) -> "ModinDataframe":
+        """
+        Allow the user to specify a type or set of types by which to filter the columns.
+
+        Parameters
+        ----------
+        types : list of hashables
+            The types to filter columns by.
+
+        Returns
+        -------
+        ModinDataframe
+             A new ModinDataframe from the filter provided.
+        """
+        pass
+
+    @abstractmethod
+    def map(
+        self,
+        function: Callable,
+        axis: Optional[int] = None,
+        dtypes: Optional[str] = None,
+    ) -> "ModinDataframe":
+        """
+        Apply a user-defined function row- wise (or column-wise if axis=1).
+
+        Parameters
+        ----------
+        function : callable
+            The function to map across the dataframe.
```

Comment:
Please add an estimated signature of the passed function here and everywhere else. Example:
```
function : calable(Series) -> Series
    The function that map across the dataframe.
```

Reply:
We probably don't want `Series` here, but `row` or `column`, agree. `Series` is too pandas-like and this level in the architecture should not be tied to pandas implementation.